### DISCUSSION TITLE
Retroactively updating a custom field for previously created assets using script

### DISCUSSION THREAD

**Discussion by Dolanap-sd** [snipe/snipe-it#15762] [Q&A]
Hello all,

I am currently tackling the challenge of adding a custom field, "Last Serviced Date," to all previously deployed assets in my org.

I have been researching how best to do this and wanted to get feedback/insight from some brilliant minds at the SnipeIT Github.

My initial thoughts were to create a script that called all the Assets of a particular type and stored the data surrounding those devices as a JSON value. The script would then reference the assets' creation date and use that to update the Last Serviced Date with a .patch command. 

I've encountered many errors when going this route, and I wanted to see if someone has any advice or experience doing something like this.

Look forward to hearing back from you all!

**Comment by snipe** [comment]
Hi @Dolanap-sd - thanks for reaching out. I'm not sure how updating the values for a custom field would be any different than a normal PATCH/PUT request. Can you clarify that for us so we can better direct you?

**Reply by Dolanap-sd** [reply]
Hi, @snipe - thanks for getting back to me. I will attach a preview of my script to clarify what I am attempting to do.

I am updating the value of the custom field in my script with a PATCH request. 

My logic in using a custom field is to have the date we last serviced the device easily visible and referencable somewhere. 

# Function to get all deployed Mac devices
def get_deployed_mac_devices():
    # Define the headers for the API request, including authorization and expected response format
    headers = {
        'Authorization': f'Bearer {API_TOKEN}',  # Use API token for authorization
        'Accept': 'application/json'              # Request JSON response
    }
    # Construct the URL to query hardware with a specific search term for MacBook Pro models
    url = f'{BASE_URL}/hardware?search=MacBook Pro (16-inch, M2)'
    
    # Send a GET request to the API
    response = requests.get(url, headers=headers)
    
    # Check if the request was successful
    if response.status_code == 200:
        # If successful, return the 'rows' from the JSON response
        return response.json()['rows']
    else:
        # If not successful, print the error message and return an empty list
        print(f"Failed to fetch devices: {response.status_code} - {response.text}")
        return []

# Function to update the Last Service Date for a specific asset
def update_last_service_date(asset_id, creation_date):
    # Define the headers for the PATCH request, including authorization and content type
    headers = {
        'Authorization': f'Bearer {API_TOKEN}',  # Use API token for authorization
        'Content-Type': 'application/json'         # Indicate the payload is in JSON format
    }
    # Construct the URL to update the specific asset by its ID
    url = f'{BASE_URL}/hardware/{asset_id}'
    
    # Prepare the payload with the new Last Service Date in the custom fields
    payload = {
        'custom_fields': {
            CUSTOM_FIELD_ID: creation_date  # Use the provided creation date for the update
        }
    }
    
    # Send a PATCH request to update the asset
    response = requests.patch(url, headers=headers, data=json.dumps(payload))
    
    # Check if the update was successful
    if response.status_code == 200:
        # If successful, print confirmation message
        print(f"Updated asset {asset_id} with Last Service Date: {creation_date}")
        return True
    else:
        # If not successful, print the error message
        print(f"Failed to update asset {asset_id}: {response.status_code} - {response.text}")
        return False

# Function to get the asset tags of updated devices
def get_asset_tags(asset_ids):
    # Define the headers for the GET request, including authorization and expected response format
    headers = {
        'Authorization': f'Bearer {API_TOKEN}',  # Use API token for authorization
        'Accept': 'application/json'              # Request JSON response
    }
    # Initialize a list to store asset tags
    asset_tags = []
    
    # Loop through each asset ID to retrieve their tags
    for asset_id in asset_ids:
        # Construct the URL to get details of the specific asset
        url = f'{BASE_URL}/hardware/{asset_id}'
        # Send a GET request to the API
        response = requests.get(url, headers=headers)
        
        # Check if the request was successful
        if response.status_code == 200:
            # If successful, extract the asset tag from the response and add it to the list
            asset_data = response.json()
            asset_tags.append(asset_data['asset_tag'])
        else:
            # If not successful, print the error message for the specific asset ID
            print(f"Failed to fetch asset tag for asset {asset_id}: {response.status_code} - {response.text}")
    
    # Return the list of asset tags
    return asset_tags

def main():
    # Fetch the deployed Mac devices
    devices = get_deployed_mac_devices()
    updated_asset_ids = []  # List to keep track of updated asset IDs
    
    # Loop through each device to update the Last Service Date
    for device in devices:
        asset_id = device['id']           # Get the asset ID from the device information
        creation_date = device['created_at']  # Use the asset's creation date
        if update_last_service_date(asset_id, creation_date):
            updated_asset_ids.append(asset_id)  # Add to the list if the update was successful

    # Fetch and print asset tags of updated assets
    asset_tags = get_asset_tags(updated_asset_ids)
    print("Updated Asset Tags:")
    for tag in asset_tags:
        print(tag)  # Print each updated asset tag

# Entry point of the script
if __name__ == '__main__':
    main()  # Execute the main function
